### PR TITLE
feat: include code duplication measurement

### PR DESCRIPTION
**Pull request by anyulled**
## What does this PR do?

Include code duplication measurement.

## Type of change

- [x] New check
- [x] Bug fix
- [x] Refactor
- [x] Documentation
- [x] CI/tooling

## Checklist

- [x] Tests added/updated (pass + fail cases for new checks)
- [x] `ruff check` and `ruff format` pass
- [x] `mypy src/` passes
- [x] Self-assessment score doesn't regress (run `ai-harness-scorecard assess .`)
- [x] ARCHITECTURE.md updated if module structure changed

## Testing

Measured with [Java repo](https://github.com/anyulled/superhero-battle-arena)

## Summary by CodeRabbit

- **New Features**
  - Added a code duplication stability check to Testing & Stability with duplication coverage scoring and CI-based remediation guidance.
- **Bug Fixes**
  - Improved CI parsing to treat blocking commands separately (including blocking `uses` inputs) and to correctly interpret `continue-on-error` values.
  - Updated blocking-command matching to prefer the blocking-specific command set when available.
- **Tests**
  - Expanded code duplication and CI parser coverage, including edge cases around normalization and lexical-block filtering.
- **Documentation**
  - Updated scorecard and README details to reflect 32 checks and refreshed the assessed timestamp.
- **Chores**
  - Renamespaced multiple check identifiers under category-specific prefixes (e.g., `testing.*`, `constraints.*`, `review.*`, etc.).

### CODE REVIEW

File: src/ai_harness_scorecard/ci_parser.py [hunk 1]
```diff
@@ -152,26 +153,44 @@ def _is_github_scheduled(data: dict[str, Any]) -> bool:
...
+        if isinstance(step_uses, str):
+            commands.append(f"uses: {step_uses}")
+            if not step_allows_failure:
+                blocking_commands.append(_format_github_action(step_uses, step.get("with")))
 
     return CIJob(
         name=name,
         commands=commands,
+        blocking_commands=blocking_commands,
         allow_failure=bool(data.get("continue-on-error", False)),
```
[the remark this answers is not in the document]
Comment (markmishaev76) [reply src/ai_harness_scorecard/ci_parser.py]: Confirmed, and it is worse than "minor" because it regresses two checks that already exist on `main`.

Reproduced against this branch. Given this workflow:

```yaml
jobs:
  lint:
    steps:
      - run: ruff check .
        continue-on-error: ${{ github.event_name == 'schedule' }}
      - run: pip-audit
        continue-on-error: ${{ github.event_name == 'schedule' }}
```

Parser output on this branch:

```
a_quoted_false   allow_failure=False  blocking=[]      # continue-on-error: "false"
b_expression     allow_failure=False  blocking=[]      # continue-on-error: ${{ ... }}
c_plain_false    allow_failure=False  blocking=['jscpd --threshold 5 src']
```

Both the quoted string and the expression are truthy under `bool()`, so the step is dropped from `blocking_commands`. Because `ci_has_blocking_command` now prefers `blocking_commands`, that propagates to existing checks:

| check | main | this branch |
|---|---|---|
| `linter_enforcement` | 4.0/4.0 "Blocking linter found in CI" | 2.0/4.0 "may not be blocking" |
| `dependency_auditing` | 4.0/4.0 "Blocking dependency audit in CI" | 2.0/4.0 "set to allow_failure" |

Expression-valued `continue-on-error` is common for "advisory on schedule, blocking on PR" setups, so this will silently halve the score of repos that are in fact enforcing the gate.

The step-level tracking itself is a genuine improvement (before this PR, a `continue-on-error: true` step still counted as blocking), so the fix is just the coercion. Something like:

```python
def _allows_failure(value: object) -> bool:
    """Only an explicit boolean true allows failure; expressions stay blocking."""
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        return value.strip().lower() == "true"
    return False
```

Applied to both the step lookup and `allow_failure=`. Note the job-level `bool(data.get("continue-on-error", False))` has the same flaw on `main` today, so fixing it in one place covers both. Worth a test for the `${{ ... }}` case since that is the one that shows up in real workflows.

File: tests/test_ci_parser.py [hunk 2]
```diff
@@ -41,6 +41,28 @@ def test_parse_github_actions_with_uses(tmp_path: Path) -> None:
     assert "./mvnw audit" in job.commands
 
 
+def test_parse_github_actions_tracks_blocking_steps_and_action_inputs(tmp_path: Path) -> None:
```
[the remark this answers is not in the document]
Comment (markmishaev76) [reply tests/test_ci_parser.py]: Disagreeing with this one, no change needed.

The `test_<check_id>_pass` / `_fail` convention in AGENTS.md applies to *check* tests, where `<check_id>` is an actual check ID. This is a parser test, and it matches the existing convention in this file:

```
tests/test_ci_parser.py
13: def test_parse_github_actions_with_uses
44: def test_parse_github_actions_schedule
64: def test_parse_github_actions_reusable_workflow_job
```

None of these carry a `_pass`/`_fail` suffix, and there is no `parse_github_actions` check ID to key off. Renaming this one test would make it the odd one out.

For what it's worth, the check tests this PR adds in `tests/test_code_duplication.py` *do* follow the convention correctly (`test_testing_code_duplication_pass`, `test_testing_code_duplication_fail`), which is where the rule actually applies.

**Review by markmishaev76 — COMMENTED** [review]
Reviewed this locally on the PR head. Verification first, then one thing I think should be fixed before merge and three notes that can be follow-ups.

**Verified locally** (Python 3.14, project `.venv`): 148 tests pass, `ruff check` and `ruff format --check` clean across 27 files, `mypy --strict` clean across 18 source files. The check is genuinely deterministic (repeat runs are byte-identical), and it is well tested, including a hypothesis invariant test. Performance is fine: on a 46,792-line Go repo the check runs in 0.72s at 20 MB peak, scaling roughly linearly, so there is no need for a file-size cap.

The additive scoring interaction is also correct. `CategoryScore.max_points` is computed as `sum(c.max_points for c in self.checks)` and the 25% weight comes from `CATEGORY_CONFIG`, so adding 3 points to Testing dilutes the other checks in the category without disturbing cross-category weighting. That matches ADR 0002.

---

### Should be fixed before merge: `continue-on-error` coercion regresses two existing checks

Details and a suggested fix are in [the thread on `ci_parser.py`](https://github.com/markmishaev76/ai-harness-scorecard/pull/52#discussion_r3668580715). Short version: `bool(step.get("continue-on-error", False))` treats the string `"false"` and any `${{ ... }}` expression as truthy, so those steps get dropped from `blocking_commands`. Since `ci_has_blocking_command` now prefers `blocking_commands`, `linter_enforcement` and `dependency_auditing` both drop from 4.0/4.0 to 2.0/4.0 on a workflow that uses `continue-on-error: ${{ github.event_name == 'schedule' }}`. That pattern is common for advisory-on-schedule gates, so real repos would silently lose points for gates they are actually enforcing.

The step-level tracking itself is a real improvement and I would like to keep it. Only the coercion needs to change.

---

### Follow-up 1: the metric counts import blocks and brace runs as duplication

`_normalized_lines` strips comments and blank lines but keeps imports and punctuation-only lines, so any 5 consecutive identical lines count. On a real Go codebase the most-repeated "duplicate" blocks are import lists and closing braces:

```
--- repeated in 64 places ---
    "github.com/MakeNowJust/heredoc/v2"
    "github.com/spf13/cobra"
    gitlab "gitlab.com/gitlab-org/api/client-go/v2"
    "gitlab.com/gitlab-org/cli/internal/api"
    "gitlab.com/gitlab-org/cli/internal/cmdutils"

--- repeated in 33 places ---
    return err
    }
    return opts.run(cmd.Context())
    },
    }
```

Measured effect on the reported number:

| repo | as implemented | ignoring imports and punctuation-only lines |
|---|---|---|
| gitlab-org/cli | 25.29% (11834/46792) | 13.99% (4535/32410) |
| model-provenance-kit | 9.74% (569/5844) | 8.45% (421/4982) |

So roughly 45% of the reported duplication on that Go repo is import blocks and brace runs. This matters because `source` cites jscpd, whose default `--min-tokens 50` would not flag a 5-line import block at all, and because the score bands are tight (2.0 at ≤5%, 1.0 at ≤10%). A healthy repo can land in the 0-point band on boilerplate alone. Worth either skipping import/`use`/`require` lines and punctuation-only lines, or gating a window on a minimum token or character count rather than a raw line count.

### Follow-up 2: reported block ranges overstate what is actually duplicated

`_largest_blocks` merges any run of consecutive duplicated indexes into one block, even when the constituent windows matched different files. On the same repo the remediation reads:

> Refactor the largest duplicate blocks: `internal/commands/token/revoke/revoke.go:96-150`, `internal/commands/variable/set/set.go:85-135`

Those two spans are 55 and 51 source lines and are not copies of each other. The genuinely shared code between them is about 10 lines (a `baseRepo` / `apiClient` preamble); the rest is adjacent unrelated small matches chained together. Since remediation is the actionable output, pointing at a 55-line span without naming its counterpart sends people to the wrong place. Recording the matching partner location per block would fix both problems.

### Follow-up 3: `check_id` convention is inconsistent with the other 31 checks

`testing.code_duplication` follows AGENTS.md, but all 31 existing checks use bare snake_case (`mutation_testing`, `property_based_testing`, `linter_enforcement`). `check_id` is emitted in the JSON reporter, so this introduces a single dotted ID into otherwise-flat output. The new ID is the *correct* one per AGENTS.md, so my preference is to keep it and migrate the rest separately rather than regress this one, but it should be a deliberate call rather than an accident. Note #50 makes the same choice with `documentation.harness_docs`.

---

### Merge coordination

#50 also adds a check and also rewrites the README from "31 checks" to "32 checks" and bumps its own category row. Whichever of the two merges second needs to say **33** and keep both category rows. This PR also edits the `Assessed` line in `scorecard-report.md`, which collides with #44.

**Comment by anyulled** [comment]
Thanks for the detailed review. I made these decisions and changes:

* Fixed `continue-on-error` parsing. Only the boolean value true makes a job or step non-blocking. The string “false” and dynamic expressions stay blocking.
* We now require both 5 lines and 50 lexical tokens for a duplicate window. This matches the **jscpd** default and avoids scoring short import lists and brace-heavy code as duplicate blocks.
* Remediation now shows both sides of a real duplicate match. It no longer joins separate nearby matches into one larger range.
* Kept `testing.code_duplication`. We also changed the prior check IDs to the same `category.snake_case` format, as required by AGENTS.md.
* The parser test name now follows the project test naming rule.

I didn't change the check count for PR #50. That needs merge-order coordination because both PRs add a check. If that is merged first, I'll rebase and increase the count.
